src/fsharp/FSharp.Core/reflect.fs

```diff
@@ -404,8 +404,10 @@ module internal Impl =
                 | _ -> invalidArg "tys" (SR.GetString(SR.invalidTupleTypes))
 
             let tables = if isStruct then valueTupleTypes else refTupleTypes
-            match tables.TryGetValue(asm) with
+            match lock dictionaryLock (fun () -> tables.TryGetValue(asm)) with
```

Comment:
Oh it makes sense what's is being attempted, of course, it is just not safe, because the Add can concurrently run on another thread and screw with internal dictionary data structures. The original Hashtable class supported that kind of concurrency, but that guarantee wasn't continued when the generic Dictionary was released. At least as far as my memory serves me. So I think it is an issue, but if you can point me to official documentation that says I'm wrong then I can be suitable enlightened.